### DISCUSSION TITLE
How to have CSS transitions on refreshable elements ?

### DISCUSSION THREAD

**Discussion by astariul** [zauberzeug/nicegui#4149] [Q&A] [answered]
### Question

I'm using [refreshable elements](https://nicegui.io/documentation/refreshable#refreshable_ui_functions) to set the opacity of another element in my page.

I'd like to add CSS transitions (for example [`transition-opacity`](https://tailwindcss.com/docs/transition-property#controlling-transitioned-properties) from Tailwind). I added the class, but it does nothing.

Here is a minimal reproducible code example : 

```python
from nicegui import ui


@ui.refreshable
def banner(opacity: bool = False):
    with ui.element("div").classes(f"inline-block opacity-{100 if opacity else 0} transition-opacity w-full"):
        ui.label("Example")


with ui.element("div").on("mouseover", lambda: banner.refresh(True)).on("mouseout", lambda: banner.refresh(False)):
    ui.label("Hover")
banner()

ui.run()
```

Here is the behavior of this code : 

![ezgif-2-d1e471e5cf](https://github.com/user-attachments/assets/78239486-c684-471d-8aac-962a73de0d43)

As you can see, there is no transition, even though I added `transition-opacity` class.

---

As far as I understand, this behavior comes from the fact that refreshable elements are deleted and recreated. So it appears to be impossible to have CSS transitions on refreshable elements, or did I miss something ?

Is there any work-around to have transition on refreshable elements ?

**Comment by rodja** [comment, marked as answer]
You are right @astariul. The refreshable simply deletes all elements and recreates them. There is no way I'm aware which would make elements inside a refreshable adhere to the `transition-opacity` class right now. Maybe you could hack the `refreshable` implementation by first hiding the old elements before deleting and then showing the new ones...

**Comment by astariul** [comment]
As @rodja pointed out, it's not possible to achieve with a "refreshable".

In my case though, I just needed to modify the style of the elements (the elements themselves stay the same), so a "refreshable" is overkill.

Instead I can just dynamically modify the classes of the elements accordingly : 

```python
from nicegui import ui


with ui.element("div") as d1:
    ui.label("Hover")

with ui.element("div").classes(f"inline-block opacity-0 transition-opacity w-full") as d2:
    ui.label("Example")

d1.on("mouseover", lambda: d2.classes("opacity-100", remove="opacity-0")).on("mouseout", lambda: d2.classes("opacity-0", remove="opacity-100"))

ui.run()
```

This code works appropriately, with the expected CSS transition : 

![ezgif-5-447d3c13f6](https://github.com/user-attachments/assets/b972f4ef-97db-4f49-ac00-6d313d192e2b)


Nicegui is awesome !

**Reply by falkoschindler** [reply]
Great! 😊

Note that, since NiceGUI 2.7, you can toggle classes:
```py
ui.label("Hover") \
    .on("mouseover", lambda: label.classes(toggle="opacity-0")) \
    .on("mouseout", lambda: label.classes(toggle="opacity-0"))
label = ui.label("Example").classes("opacity-0 transition-opacity")
```

**Reply by astariul** [reply]
Thanks @falkoschindler, I didn't know about toggle.

Your code snippet made me wonder : is it possible to define a single callback for several events ? Something like : 

```python
ui.label("Hover") \
    .on(["mouseover", "mouseout"], lambda: label.classes(toggle="opacity-0"))
label = ui.label("Example").classes("opacity-0 transition-opacity")
```

From the documentation it doesn't seem so, but I'm asking just in case :)

**Reply by falkoschindler** [reply]
No, registering event handlers for multiple events at once is currently not implemented. But you can avoid move of the copy&paste by moving the handler into a proper function rather than a lambda expression:
```py
def toggle_opacity() -> None:
    label.classes(toggle="opacity-0")

ui.label("Hover") \
    .on("mouseover", toggle_opacity) \
    .on("mouseout", toggle_opacity)
label = ui.label("Example").classes("opacity-0 transition-opacity")
```

**Reply by astariul** [reply]
Thanks !
